File: src/api/freeSessions.ts
/**
 * GRACE FREE daily session service (Milestone 13).
 *
 * Enforces the free plan on the server — the CLI never stores or trusts any
 * session state, so restarting it or deleting local files can never reset the
 * daily quota:
 *
 *   * 6 sessions per user per day        (ZEESH_SESSIONS_PER_DAY, default 6)
 *   * 60 minutes per session             (ZEESH_SESSION_DURATION_MINUTES, default 60)
 *   * 6 hours / day max                  (sessionsPerDay × sessionDuration)
 *   * day boundary = 00:00 UTC           (server-authoritative, timezone-independent)
 *
 * State lives in Neon `free_sessions` (db/migrations/004_free_sessions.sql):
 * one row per session with a UNIQUE (user_id, day, session_number) constraint,
 * which is what makes the "start the next session" step race-safe (concurrent
 * requests that both pick the same number collide → the loser retries with the
 * new MAX).
 *
 * API:
 *   - `getState(userId)`            read-only summary for GET /api/usage.
 *   - `ensureActiveSession(userId)` the authoritative gate for /api/provider:
 *       · active session   → serve the request inside it,
 *       · expired / none   → auto-start the next session if quota remains
 *                            ("automatically move the user to the next
 *                            session"), otherwise refuse with
 *                            `{ code: 'daily_limit_exhausted' }`.
 */
import type { Db, Row } from './db.ts';

/** Default free-plan limits (overridable per deployment via env, like rate limits). */
export const DEFAULT_SESSIONS_PER_DAY = 6;
export const DEFAULT_SESSION_DURATION_MS = 60 * 60 * 1000; // 60 minutes

/** A row of `free_sessions` as read by this service. */
export interface FreeSessionRow {
  id: string;
  user_id: string;
  /** UTC day bucket (YYYY-MM-DD) the session belongs to. */
  day: string;
  session_number: number;
  /** ISO timestamps. */
  started_at: string;
  expires_at: string;
  ended_at: string | null;
}

/**
 * The daily free-plan summary. Mirrored on the CLI side as
 * `DailySessionState` in src/auth/client.ts — keep both in sync.
 */
export interface DailySessionState {
  sessionsUsed: number;
  sessionsRemaining: number;
  /** Number of the active session, or null when none is active (expired/none). */
  currentSession: number | null;
  sessionStartedAt: string | null;
  sessionExpiresAt: string | null;
  /** Wall-clock seconds consumed today (expired sessions count in full). */
  dailyUsedSeconds: number;
  dailyLimitSeconds: number;
}

/** Result of the /api/provider gate. */
export type SessionGate =
  | { ok: true; state: DailySessionState; startedNew: boolean }
  | { ok: false; status: number; code: 'daily_limit_exhausted'; error: string; state: DailySessionState };

export interface FreeSessionServiceOptions {
  /** Injectable clock (tests). Defaults to `new Date()`. */
  now?: () => Date;
}

const MAX_START_ATTEMPTS = 10;

/**
 * The UTC date bucket a session belongs to (YYYY-MM-DD). The day boundary is
 * deliberately UTC — the server's authoritative day is the same for every
 * user regardless of timezone, so quota math is unambiguous.
 */
export function utcDay(d: Date): string {
  return d.toISOString().slice(0, 10);
}

/** Seconds until the next UTC midnight (used for the 429 Retry-After header). */
export function secondsUntilUtcMidnight(now: Date = new Date()): number {
  const next = new Date(now);
  next.setUTCHours(24, 0, 0, 0);
  return Math.max(1, Math.ceil((next.getTime() - now.getTime()) / 1000));
}

/** Read an env override as a positive integer, falling back to `fallback`. */
function envPositiveInt(name: string, fallback: number): number {
  const raw = Number(process.env[name]);
  return Number.isFinite(raw) && raw > 0 ? Math.floor(raw) : fallback;
}

export class FreeSessionService {
  private readonly db: Db;
  private readonly options: FreeSessionServiceOptions;

  constructor(db: Db, options: FreeSessionServiceOptions = {}) {
    this.db = db;
    this.options = options;
  }

  private now(): Date {
    return this.options.now?.() ?? new Date();
  }

  private get sessionsPerDay(): number {
    return envPositiveInt('ZEESH_SESSIONS_PER_DAY', DEFAULT_SESSIONS_PER_DAY);
  }

  private get sessionDurationMs(): number {
    return envPositiveInt('ZEESH_SESSION_DURATION_MINUTES', DEFAULT_SESSION_DURATION_MS / 60_000) * 60_000;
  }

  // -------------------------------------------------------------------------
  // Read-only state (GET /api/usage) — never mutates.
  // -------------------------------------------------------------------------

  /** Daily free-plan summary for a user, computed from Neon. */
  async getState(userId: string): Promise<DailySessionState> {
    const now = this.now();
    const rows = await this.sessionRows(userId, utcDay(now));
    return this.computeState(rows, now);
  }

  /** Authoritative gate for /api/provider (may auto-start the next session). */
  async ensureActiveSession(userId: string): Promise<SessionGate> {
    const now = this.now();
    const day = utcDay(now);
    const limit = this.sessionsPerDay;

    // Bounded retries: concurrent requests that pick the same session_number
    // collide on the UNIQUE constraint (SQLSTATE 23505) and re-run against the
    // new MAX. With limit slots per day and eager re-reads, a request never
    // needs more than a couple of attempts.
    for (let attempt = 0; attempt < MAX_START_ATTEMPTS; attempt += 1) {
      const rows = await this.sessionRows(userId, day);
      const last = rows[rows.length - 1];

      if (last && isActive(last, now)) {
        // Current session is still live — serve inside it.
        return { ok: true, state: this.computeState(rows, now), startedNew: false };
      }

      if (rows.length >= limit) {
        // All sessions for today are used up (and the last one is expired).
        if (last) await this.markEnded(last); // lazy end: session expired
        return {
          ok: false,
          status: 429,
          code: 'daily_limit_exhausted',
          error:
            `You have used all ${limit} free sessions for today (${Math.round((limit * this.sessionDurationMs) / 3_600_000)}h max). ` +
            'New sessions unlock at 00:00 UTC. Thanks for using GRACE FREE.',
          // The state rides along so the rejection is self-describing (the
          // CLI can render "Session 6/6" from the 429 itself).
          state: this.computeState(rows, now),
        };
      }

      // Expired session (or none yet) + quota remains → auto-start the next.
      const sessionNumber = rows.length + 1;
      const startedAt = now;
      const expiresAt = new Date(now.getTime() + this.sessionDurationMs);
      try {
        const inserted = await this.db(
          `INSERT INTO free_sessions (user_id, day, session_number, started_at, expires_at)
           VALUES ($1, $2, $3, $4, $5)
           RETURNING id, user_id, day, session_number, started_at, expires_at, ended_at`,
          [userId, day, sessionNumber, startedAt.toISOString(), expiresAt.toISOString()],
        );
        const row = inserted[0];
        if (!row) throw new Error('Could not start a free session.');
        // The session we just replaced expired — mark it ended (lazy expiry).
        if (last) await this.markEnded(last);
        const all = [...rows, toFreeSessionRow(row)];
        return { ok: true, state: this.computeState(all, now), startedNew: true };
      } catch (err) {
        // Unique violation → another request started this number first; retry.
        if (isUniqueViolation(err)) continue;
        throw err;
      }
    }

    // Retry budget exhausted — most likely the day filled up under contention.
    const rows = await this.sessionRows(userId, day);
    if (rows.length >= limit) {
      return {
        ok: false,
        status: 429,
        code: 'daily_limit_exhausted',
        error: `You have used all ${limit} free sessions for today. New sessions unlock at 00:00 UTC.`,
        state: this.computeState(rows, now),
      };
    }
    throw new Error('Could not start a free session.');
  }

  // -------------------------------------------------------------------------
  // Internals
  // -------------------------------------------------------------------------

  private async sessionRows(userId: string, day: string): Promise<FreeSessionRow[]> {
    const rows = await this.db(
      `SELECT id, user_id, day, session_number, started_at, expires_at, ended_at
         FROM free_sessions
        WHERE user_id = $1 AND day = $2
        ORDER BY session_number ASC`,
      [userId, day],
    );
    return rows.map(toFreeSessionRow);
  }

  /** Mark a session ended (idempotent) — the DB stays an explicit record. */
  private async markEnded(row: FreeSessionRow): Promise<void> {
    await this.db('UPDATE free_sessions SET ended_at = expires_at WHERE id = $1 AND ended_at IS NULL', [row.id]);
  }

  /** Derive the daily summary from the day's rows. */
  private computeState(rows: FreeSessionRow[], now: Date): DailySessionState {
    const limit = this.sessionsPerDay;
    const durationMs = this.sessionDurationMs;
    const nowMs = now.getTime();

    let usedSeconds = 0;
    for (const row of rows) {
      const start = new Date(row.started_at).getTime();
      const end = new Date(row.expires_at).getTime();
      // Elapsed so far, capped at the session's own expiry (60 min max each).
      const elapsedMs = Math.min(nowMs, end) - start;
      if (elapsedMs > 0) usedSeconds += elapsedMs;
    }
    // Never exceed the daily cap (guard against clock skew / overlapping rows).
    const dailyLimitSeconds = Math.round((limit * durationMs) / 1000);
    const dailyUsedSeconds = Math.min(Math.round(usedSeconds / 1000), dailyLimitSeconds);

    const last = rows[rows.length - 1];
    const active = last !== undefined && isActive(last, now);

    return {
      sessionsUsed: rows.length,
      sessionsRemaining: Math.max(0, limit - rows.length),
      currentSession: active ? last.session_number : null,
      sessionStartedAt: active ? last.started_at : null,
      sessionExpiresAt: active ? last.expires_at : null,
      dailyUsedSeconds,
      dailyLimitSeconds,
    };
  }
}

/** A session is active while now < expires_at. */
function isActive(row: FreeSessionRow, now: Date): boolean {
  const expiresAt = new Date(row.expires_at).getTime();
  return !Number.isNaN(expiresAt) && expiresAt > now.getTime();
}

/** Postgres UNIQUE violation (SQLSTATE 23505). The memory test db mirrors it. */
function isUniqueViolation(err: unknown): boolean {
  return (err as { code?: string } | null)?.code === '23505';
}

/** Map a raw db row to the typed shape (numbers are returned as such by Neon). */
function toFreeSessionRow(row: Row): FreeSessionRow {
  return {
    id: String(row.id),
    user_id: String(row.user_id),
    day: String(row.day),
    session_number: Number(row.session_number),
    started_at: String(row.started_at),
    expires_at: String(row.expires_at),
    ended_at: row.ended_at == null ? null : String(row.ended_at),
  };
}

File: grace/server/free_sessions.py
"""GRACE FREE daily session service (port of src/api/freeSessions.ts).

Enforces the free plan on the server — the CLI never stores or trusts any
session state, so restarting it or deleting local files can never reset the
daily quota:

  * 3 sessions per user per day        (ZEESH_SESSIONS_PER_DAY, default 3)
  * 60 minutes per session             (ZEESH_SESSION_DURATION_MINUTES, default 60)
  * 3 hours / day max                  (sessionsPerDay × sessionDuration)
  * day boundary = 00:00 UTC           (server-authoritative, timezone-independent)

State lives in Neon `free_sessions` (db/migrations/004_free_sessions.sql):
one row per session with a UNIQUE (user_id, day, session_number) constraint,
which is what makes the "start the next session" step race-safe (concurrent
requests that both pick the same number collide → the loser retries with the
new MAX).

API:
  - get_state(userId)            read-only summary for GET /api/usage.
  - ensure_active_session(userId) the authoritative gate for /api/provider:
      · active session   → serve the request inside it,
      · expired / none   → auto-start the next session if quota remains
                           ("automatically move the user to the next
                           session"), otherwise refuse with
                           `{ code: 'daily_limit_exhausted' }`.
"""

import math
import os
from datetime import datetime, timezone

from grace.server.db import Db

DEFAULT_SESSIONS_PER_DAY = 3
DEFAULT_SESSION_DURATION_MS = 60 * 60 * 1000  # 60 minutes
MAX_START_ATTEMPTS = 10


def utc_day(d: datetime) -> str:
    """The UTC date bucket a session belongs to (YYYY-MM-DD). The day boundary
    is deliberately UTC — the server's authoritative day is the same for every
    user regardless of timezone, so quota math is unambiguous."""
    return d.astimezone(timezone.utc).strftime("%Y-%m-%d")


def seconds_until_utc_midnight(now: datetime | None = None) -> int:
    """Seconds until the next UTC midnight (used for the 429 Retry-After header)."""
    now = now or datetime.now(timezone.utc)
    now_ms = int(now.astimezone(timezone.utc).timestamp() * 1000)
    next_midnight = datetime(now.year, now.month, now.day, tzinfo=timezone.utc).timestamp() * 1000 + 24 * 60 * 60 * 1000
    return max(1, math.ceil((next_midnight - now_ms) / 1000))


def env_positive_int(name: str, fallback: int) -> int:
    """Read an env override as a positive integer, falling back to `fallback`."""
    try:
        raw = float(os.environ.get(name) or "")
    except (TypeError, ValueError):
        return fallback
    return math.floor(raw) if math.isfinite(raw) and raw > 0 else fallback


def _ts_round(x: float) -> int:
    """Math.round semantics (half away from zero) — Python's round() is banker's."""
    return int(math.floor(x + 0.5))


class FreeSessionService:
    def __init__(self, db: Db, options: dict | None = None) -> None:
        self.db = db
        self.options = options or {}

    def now(self) -> datetime:
        return self.options.get("now", lambda: datetime.now(timezone.utc))()

    @property
    def sessions_per_day(self) -> int:
        return env_positive_int("ZEESH_SESSIONS_PER_DAY", DEFAULT_SESSIONS_PER_DAY)

    @property
    def session_duration_ms(self) -> int:
        return env_positive_int("ZEESH_SESSION_DURATION_MINUTES", DEFAULT_SESSION_DURATION_MS // 60_000) * 60_000

    # -------------------------------------------------------------------------
    # Read-only state (GET /api/usage) — never mutates.
    # -------------------------------------------------------------------------

    def get_state(self, user_id: str) -> dict:
        """Daily free-plan summary for a user, computed from Neon."""
        now = self.now()
        rows = self._session_rows(user_id, utc_day(now))
        return self._compute_state(rows, now)

    def active_session_row(self, user_id: str) -> dict | None:
        """The user's currently ACTIVE session row, or None (read-only — never
        starts or ends anything). Used by GET /api/session/status."""
        now = self.now()
        rows = self._session_rows(user_id, utc_day(now))
        last = rows[-1] if rows else None
        return last if last is not None and _is_active(last, now) else None

    def last_session_row(self, user_id: str) -> dict | None:
        """The day's most recent session row, or None when the user has no
        session yet today (read-only). Lets the status endpoint distinguish an
        explicitly ended session (ended_at < expires_at) from one that expired
        naturally (ended_at == expires_at) and from no session at all."""
        rows = self._session_rows(user_id, utc_day(self.now()))
        return rows[-1] if rows else None

    def end_active_session(self, user_id: str) -> dict:
        """Explicitly end the user's active session (POST /api/session/end).
        Marks the active row ended now; never starts a replacement — ending is
        an explicit action, quota stays as-is. Idempotent: no active session is
        a no-op (the expired row is lazily marked ended)."""
        now = self.now()
        day = utc_day(now)
        rows = self._session_rows(user_id, day)
        last = rows[-1] if rows else None
        if last is not None:
            if _is_active(last, now):
                self.db(
                    "UPDATE free_sessions SET ended_at = $2 WHERE id = $1 AND ended_at IS NULL",
                    [last["id"], now.astimezone(timezone.utc).isoformat()],
                )
            else:
                self._mark_ended(last)  # lazy end of an already-expired session
        after = self._session_rows(user_id, day)
        return self._compute_state(after, now)

    def ensure_active_session(self, user_id: str) -> dict:
        """Authoritative gate for /api/provider (may auto-start the next session)."""
        now = self.now()
        day = utc_day(now)
        limit = self.sessions_per_day

        for _ in range(MAX_START_ATTEMPTS):
            rows = self._session_rows(user_id, day)
            last = rows[-1] if rows else None

            if last is not None and _is_active(last, now):
                # Current session is still live — serve inside it.
                return {"ok": True, "state": self._compute_state(rows, now), "startedNew": False}

            if len(rows) >= limit:
                # All sessions for today are used up (and the last one is expired).
                if last is not None:
                    self._mark_ended(last)  # lazy end: session expired
                return {
                    "ok": False,
                    "status": 429,
                    "code": "daily_limit_exhausted",
                    "error": (
                        f"You have used all {limit} free sessions for today "
                        f"({_ts_round((limit * self.session_duration_ms) / 3_600_000)}h max). "
                        "New sessions unlock at 00:00 UTC. Thanks for using GRACE FREE."
                    ),
                    "state": self._compute_state(rows, now),
                }

            # Expired session (or none yet) + quota remains → auto-start the next.
            session_number = len(rows) + 1
            started_at = now
            expires_at = started_at + _ms_timedelta(self.session_duration_ms)
            try:
                inserted = self.db(
                    "INSERT INTO free_sessions (user_id, day, session_number, started_at, expires_at)"
                    " VALUES ($1, $2, $3, $4, $5)"
                    " RETURNING id, user_id, day, session_number, started_at, expires_at, ended_at",
                    [user_id, day, session_number, _iso(started_at), _iso(expires_at)],
                )
                row = inserted[0] if inserted else None
                if not row:
                    raise RuntimeError("Could not start a free session.")
                # The session we just replaced expired — mark it ended (lazy expiry).
                if last is not None:
                    self._mark_ended(last)
                all_rows = list(rows) + [_to_free_session_row(row)]
                return {"ok": True, "state": self._compute_state(all_rows, now), "startedNew": True}
            except Exception as err:
                # Unique violation → another request started this number first; retry.
                if _is_unique_violation(err):
                    continue
                raise err

        # Retry budget exhausted — most likely the day filled up under contention.
        rows = self._session_rows(user_id, day)
        if len(rows) >= limit:
            return {
                "ok": False,
                "status": 429,
                "code": "daily_limit_exhausted",
                "error": f"You have used all {limit} free sessions for today. New sessions unlock at 00:00 UTC.",
                "state": self._compute_state(rows, now),
            }
        raise RuntimeError("Could not start a free session.")

    # -------------------------------------------------------------------------
    # Internals
    # -------------------------------------------------------------------------

    def _session_rows(self, user_id: str, day: str) -> list[dict]:
        rows = self.db(
            "SELECT id, user_id, day, session_number, started_at, expires_at, ended_at"
            "   FROM free_sessions"
            "  WHERE user_id = $1 AND day = $2"
            "  ORDER BY session_number ASC",
            [user_id, day],
        )
        return [_to_free_session_row(r) for r in rows]

    def _mark_ended(self, row: dict) -> None:
        """Mark a session ended (idempotent) — the DB stays an explicit record."""
        self.db("UPDATE free_sessions SET ended_at = expires_at WHERE id = $1 AND ended_at IS NULL", [row["id"]])

    def _compute_state(self, rows: list[dict], now: datetime) -> dict:
        limit = self.sessions_per_day
        duration_ms = self.session_duration_ms
        now_ms = int(now.astimezone(timezone.utc).timestamp() * 1000)

        used_seconds = 0
        for row in rows:
            start = _iso_ms(row["started_at"])
            end = _iso_ms(row["expires_at"])
            if start is None or end is None:
                continue  # unparseable timestamps are skipped, like NaN in TS
            # Elapsed so far, capped at the session's own expiry (60 min max each).
            elapsed_ms = min(now_ms, end) - start
            if elapsed_ms > 0:
                used_seconds += elapsed_ms
        # Never exceed the daily cap (guard against clock skew / overlapping rows).
        daily_limit_seconds = _ts_round((limit * duration_ms) / 1000)
        daily_used_seconds = min(_ts_round(used_seconds / 1000), daily_limit_seconds)

        last = rows[-1] if rows else None
        active = last is not None and _is_active(last, now)

        return {
            "sessionsUsed": len(rows),
            "sessionsRemaining": max(0, limit - len(rows)),
            "currentSession": last["session_number"] if active else None,
            "sessionStartedAt": last["started_at"] if active else None,
            "sessionExpiresAt": last["expires_at"] if active else None,
            "dailyUsedSeconds": daily_used_seconds,
            "dailyLimitSeconds": daily_limit_seconds,
        }


def _is_active(row: dict, now: datetime) -> bool:
    """A session is active while now < expires_at AND it was not ended early
    (POST /api/session/end sets ended_at — an explicitly ended session is never
    reused, even while its expiry is still in the future)."""
    expires_at = _iso_ms(row["expires_at"])
    ended_at = _iso_ms(row["ended_at"]) if row.get("ended_at") is not None else None
    now_ms = int(now.astimezone(timezone.utc).timestamp() * 1000)
    if expires_at is None:
        return False
    if ended_at is not None and ended_at <= now_ms:
        return False
    return expires_at > now_ms


def _is_unique_violation(err: Exception) -> bool:
    """Postgres UNIQUE violation (SQLSTATE 23505). The memory test db mirrors it."""
    return getattr(err, "code", None) == "23505" or getattr(err, "sqlstate", None) == "23505"


def _to_free_session_row(row: dict) -> dict:
    return {
        "id": str(row.get("id")),
        "user_id": str(row.get("user_id")),
        "day": str(row.get("day")),
        "session_number": int(row.get("session_number")),
        "started_at": str(row.get("started_at")),
        "expires_at": str(row.get("expires_at")),
        "ended_at": row.get("ended_at") if row.get("ended_at") is not None else None,
    }


def _iso(d: datetime) -> str:
    return d.astimezone(timezone.utc).isoformat()


def _ms_timedelta(ms: int):
    from datetime import timedelta

    return timedelta(milliseconds=ms)


def _iso_ms(value: str) -> int | None:
    """Parse an ISO timestamp to epoch ms, or None when unparseable."""
    try:
        return int(datetime.fromisoformat(value.replace("Z", "+00:00")).timestamp() * 1000)
    except (ValueError, TypeError):
        return None
